exonum-node/src/events/network.rs

```diff
@@ -105,135 +133,166 @@ pub struct NetworkPart {
+            .count()
     }
 
-    fn add_incoming_address(
-        &self,
-        key: &PublicKey,
-        address: &ConnectedPeerAddr,
-    ) -> mpsc::Receiver<SignedMessage> {
-        let (sender_tx, receiver_rx) = mpsc::channel::<SignedMessage>(OUTGOING_CHANNEL_SIZE);
-        self.add(key, address.clone(), sender_tx);
-        receiver_rx
+    fn count_outgoing(&self) -> usize {
+        self.peers
+            .values()
+            .filter(|entry| entry.address.is_incoming())
+            .count()
     }
 
-    fn send_message(
-        &self,
-        address: &PublicKey,
-        message: SignedMessage,
-    ) -> impl Future<Item = (), Error = failure::Error> + Send {
-        let address = *address;
-        let sender_tx = self.peers.read().unwrap();
-        let write_pool = self.clone();
-
-        if let Some(entry) = sender_tx.get(&address) {
-            let sender = &entry.sender;
-            Either::A(
-                sender
-                    .clone()
-                    .send(message)
-                    .map(drop)
-                    .or_else(move |e| {
-                        log_error(e);
-                        write_pool.remove(&address);
-                        Ok(())
-                    })
-                    .map(drop),
-            )
-        } else {
-            Either::B(future::ok(()))
-        }
+    /// Adds a peer to the connection list.
+    ///
+    /// # Return value
+    ///
+    /// Returns the receiver for outgoing messages to the peer and the connection ID.
+    fn add(
+        &mut self,
+        key: PublicKey,
+        address: ConnectedPeerAddr,
+    ) -> (mpsc::Receiver<SignedMessage>, u64) {
+        let id = self.next_connection_id;
+        let (sender, receiver_rx) = mpsc::channel(OUTGOING_CHANNEL_SIZE);
+        let entry = ConnectionPoolEntry {
+            sender,
+            address,
+            id,
+        };
+
+        self.next_connection_id += 1;
+        self.peers.insert(key, entry);
+        (receiver_rx, id)
     }
 
-    fn disconnect_with_peer(
-        &self,
-        key: &PublicKey,
-        network_tx: &mpsc::Sender<NetworkEvent>,
-    ) -> impl Future<Item = (), Error = failure::Error> + Send {
-        if self.remove(key).is_some() {
-            let send_disconnected = network_tx
-                .clone()
-                .send(NetworkEvent::PeerDisconnected(*key))
-                .map_err(|_| format_err!("can't send disconnect"))
-                .map(drop);
-            Either::A(send_disconnected)
-        } else {
-            Either::B(future::ok(()))
+    fn contains(&self, address: &PublicKey) -> bool {
+        self.peers.get(address).is_some()
+    }
+
+    /// Drops the connection to a peer. The request can be optionally filtered by the connection ID
+    /// in order to avoid issuing obsolete requests.
+    ///
+    /// # Return value
+    ///
+    /// Returns `true` if the connection with the peer was dropped.
```

Comment:
I'd add a description when we return `false`. If the connection with address doesn't exist we return `false`.